File: backend/domain_monitor/spam.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

import config_loader
from anchors import AnchorProfile
from topics import TextEvidence, count_hits

LEVEL_LOW = "Low"
LEVEL_MODERATE = "Moderate"
LEVEL_HIGH = "High"
LEVEL_VERY_HIGH = "Very High"


@dataclass
class SpamSignal:
    """One triggered rule. `detail` is what the UI shows the user verbatim."""

    code: str
    label: str
    detail: str
    points: int


@dataclass
class SpamAssessment:
    score: Optional[int] = None
    level: Optional[str] = None
    signals: list[SpamSignal] = field(default_factory=list)
    detected_categories: list[str] = field(default_factory=list)
    evaluated_rules: list[str] = field(default_factory=list)

    @property
    def has_data(self) -> bool:
        return self.score is not None


def level_for(score: int) -> str:
    levels = config_loader.scoring().get("spam_risk", {}).get("levels", {})
    if score <= int(levels.get("low_max", 20)):
        return LEVEL_LOW
    if score <= int(levels.get("moderate_max", 45)):
        return LEVEL_MODERATE
    if score <= int(levels.get("high_max", 70)):
        return LEVEL_HIGH
    return LEVEL_VERY_HIGH


def spam_keyword_set() -> set[str]:
    """Flat set of every configured spam keyword token, for anchor matching."""
    words: set[str] = set()
    for spec in config_loader.spam_categories().values():
        for keyword in spec.get("keywords", []):
            words.update(keyword.lower().split())
    return words


def detect_categories(evidence: TextEvidence) -> list[tuple[str, int, int]]:
    """Spam categories present in the domain's history.

    Returns (category, hit_count, configured_points) triples.
    """
    if evidence.is_empty:
        return []

    haystack = evidence.joined()
    found: list[tuple[str, int, int]] = []
    for name, spec in config_loader.spam_categories().items():
        hits = count_hits(haystack, spec.get("keywords", []))
        if hits > 0:
            found.append((name, hits, int(spec.get("points", 10))))
    return sorted(found, key=lambda row: -row[2])


def assess(
    domain: str,
    evidence: TextEvidence,
    anchor_profile: AnchorProfile,
    topic_switch_count: Optional[int],
    referring_domains: Optional[int],
    new_backlinks: Optional[int],
    lost_backlinks: Optional[int],
    total_backlinks: Optional[int],
) -> SpamAssessment:
    """Deterministic spam risk. Every point is traceable to one printed rule.

    Returns an assessment with score None when no input signal was available at
    all — an unknown history is not a clean history.
    """
    cfg = config_loader.scoring().get("spam_risk", {})
    signals: list[SpamSignal] = []
    evaluated: list[str] = []

    # --- historical spam content -------------------------------------------
    categories: list[str] = []
    if not evidence.is_empty:
        evaluated.append("historical_content")
        for name, hits, points in detect_categories(evidence):
            categories.append(name)
            signals.append(
                SpamSignal(
                    code=f"history_{name.lower().replace(' ', '_')}",
                    label=f"Historical {name} content",
                    detail=f"{hits} {name.lower()} keyword hit{'s' if hits != 1 else ''} in archived titles, meta or anchors",
                    points=points,
                )
            )

    # --- anchor profile -----------------------------------------------------
    if anchor_profile.has_data:
        evaluated.append("anchor_profile")

        threshold = float(cfg.get("suspicious_anchor_pct_threshold", 25))
        if anchor_profile.suspicious_pct is not None and anchor_profile.suspicious_pct >= threshold:
            signals.append(
                SpamSignal(
                    code="suspicious_anchors",
                    label="Suspicious anchor concentration",
                    detail=f"{anchor_profile.suspicious_pct}% of anchors contain spam keywords (threshold {threshold:g}%)",
                    points=int(cfg.get("suspicious_anchor_points", 20)),
                )
            )

        exact_threshold = float(cfg.get("exact_match_concentration_threshold", 35))
        if anchor_profile.exact_match_pct is not None and anchor_profile.exact_match_pct >= exact_threshold:
            signals.append(
                SpamSignal(
                    code="exact_match_anchors",
                    label="High exact-match anchor ratio",
                    detail=f"{anchor_profile.exact_match_pct}% exact-match anchors (threshold {exact_threshold:g}%)",
                    points=int(cfg.get("exact_match_points", 12)),
                )
            )

    # --- topic volatility ---------------------------------------------------
    if topic_switch_count is not None:
        evaluated.append("topic_switching")
        switch_threshold = int(cfg.get("topic_switch_threshold", 3))
        if topic_switch_count >= switch_threshold:
            points = min(
                topic_switch_count * int(cfg.get("topic_switch_points_each", 6)),
                int(cfg.get("topic_switch_points_cap", 24)),
            )
            signals.append(
                SpamSignal(
                    code="topic_switching",
                    label="Unrelated topic changes",
                    detail=f"{topic_switch_count} topic changes across the archive (threshold {switch_threshold})",
                    points=points,
                )
            )

    # --- backlink shape ----------------------------------------------------
    if total_backlinks is not None and referring_domains:
        evaluated.append("backlink_shape")
        ratio = total_backlinks / referring_domains
        spike_ratio = float(cfg.get("backlink_spike_ratio", 3.0))
        # Many links from very few domains is the classic footprint of a
        # sitewide or network link blast.
        if ratio >= spike_ratio * 25:
            signals.append(
                SpamSignal(
                    code="backlink_concentration",
                    label="Abnormal backlink concentration",
                    detail=f"{ratio:.0f} backlinks per referring domain — links come from very few sources",
                    points=int(cfg.get("backlink_spike_points", 14)),
                )
            )

    if lost_backlinks is not None and total_backlinks:
        evaluated.append("backlink_decay")
        lost_ratio = lost_backlinks / max(total_backlinks, 1)
        if lost_ratio >= float(cfg.get("lost_backlink_ratio", 0.6)):
            signals.append(
                SpamSignal(
                    code="backlink_decay",
                    label="Heavy backlink loss",
                    detail=f"{lost_backlinks:,} lost vs {total_backlinks:,} live backlinks ({lost_ratio * 100:.0f}%)",
                    points=int(cfg.get("lost_backlink_points", 8)),
                )
            )

    if not evaluated:
        # No history, no anchors, no backlinks: we genuinely cannot say.
        return SpamAssessment()

    score = min(100, sum(signal.points for signal in signals))
    return SpamAssessment(
        score=score,
        level=level_for(score),
        signals=signals,
        detected_categories=categories,
        evaluated_rules=evaluated,
    )

File: node-backend/domainMonitor/spam.js
"use strict";

// Faithful Node port of backend/domain_monitor/spam.py
const configLoader = require("./configLoader");
const { countHits } = require("./topics");

const LEVEL_LOW = "Low";
const LEVEL_MODERATE = "Moderate";
const LEVEL_HIGH = "High";
const LEVEL_VERY_HIGH = "Very High";

// --- number formatting mirrors Python f-string specs ---------------------
function _comma(n) {
  return Number(n).toLocaleString("en-US");
}
// ":g" — general format, strips trailing zeros (25.0 -> "25").
function _g(n) {
  return Number(n).toString();
}
// str(float) after round(_, 1): always one decimal place (25.0 -> "25.0").
function _pct1(n) {
  return Number(n).toFixed(1);
}
// ":.0f"
function _fixed0(n) {
  return String(Math.round(n));
}

function levelFor(score) {
  const levels = (configLoader.scoring().spam_risk || {}).levels || {};
  if (score <= parseInt(levels.low_max ?? 20, 10)) return LEVEL_LOW;
  if (score <= parseInt(levels.moderate_max ?? 45, 10)) return LEVEL_MODERATE;
  if (score <= parseInt(levels.high_max ?? 70, 10)) return LEVEL_HIGH;
  return LEVEL_VERY_HIGH;
}

/** Flat set of every configured spam keyword token, for anchor matching. */
function spamKeywordSet() {
  const words = new Set();
  for (const spec of Object.values(configLoader.spamCategories())) {
    for (const keyword of (spec && spec.keywords) || []) {
      for (const token of keyword.toLowerCase().split(/\s+/)) {
        if (token) words.add(token);
      }
    }
  }
  return words;
}

/**
 * Spam categories present in the domain's history.
 * Returns [category, hitCount, configuredPoints] triples.
 */
function detectCategories(evidence) {
  if (evidence.is_empty) return [];

  const haystack = evidence.joined();
  const found = [];
  for (const [name, spec] of Object.entries(configLoader.spamCategories())) {
    const hits = countHits(haystack, (spec && spec.keywords) || []);
    if (hits > 0) found.push([name, hits, parseInt((spec && spec.points) ?? 10, 10)]);
  }
  found.sort((a, b) => b[2] - a[2]);
  return found;
}

/**
 * Deterministic spam risk. Every point is traceable to one printed rule.
 *
 * Returns an assessment with score null when no input signal was available at
 * all — an unknown history is not a clean history.
 */
function assess({
  domain,
  evidence,
  anchor_profile,
  topic_switch_count,
  referring_domains,
  new_backlinks,
  lost_backlinks,
  total_backlinks,
} = {}) {
  const cfg = configLoader.scoring().spam_risk || {};
  const signals = [];
  const evaluated = [];

  // --- historical spam content -------------------------------------------
  const categories = [];
  if (!evidence.is_empty) {
    evaluated.push("historical_content");
    for (const [name, hits, points] of detectCategories(evidence)) {
      categories.push(name);
      signals.push({
        code: `history_${name.toLowerCase().replace(/ /g, "_")}`,
        label: `Historical ${name} content`,
        detail: `${hits} ${name.toLowerCase()} keyword hit${hits !== 1 ? "s" : ""} in archived titles, meta or anchors`,
        points,
      });
    }
  }

  // --- anchor profile -----------------------------------------------------
  if (anchor_profile.has_data) {
    evaluated.push("anchor_profile");

    const threshold = parseFloat(cfg.suspicious_anchor_pct_threshold ?? 25);
    if (anchor_profile.suspicious_pct !== null && anchor_profile.suspicious_pct >= threshold) {
      signals.push({
        code: "suspicious_anchors",
        label: "Suspicious anchor concentration",
        detail: `${_pct1(anchor_profile.suspicious_pct)}% of anchors contain spam keywords (threshold ${_g(threshold)}%)`,
        points: parseInt(cfg.suspicious_anchor_points ?? 20, 10),
      });
    }

    const exactThreshold = parseFloat(cfg.exact_match_concentration_threshold ?? 35);
    if (anchor_profile.exact_match_pct !== null && anchor_profile.exact_match_pct >= exactThreshold) {
      signals.push({
        code: "exact_match_anchors",
        label: "High exact-match anchor ratio",
        detail: `${_pct1(anchor_profile.exact_match_pct)}% exact-match anchors (threshold ${_g(exactThreshold)}%)`,
        points: parseInt(cfg.exact_match_points ?? 12, 10),
      });
    }
  }

  // --- topic volatility ---------------------------------------------------
  if (topic_switch_count !== null && topic_switch_count !== undefined) {
    evaluated.push("topic_switching");
    const switchThreshold = parseInt(cfg.topic_switch_threshold ?? 3, 10);
    if (topic_switch_count >= switchThreshold) {
      const points = Math.min(
        topic_switch_count * parseInt(cfg.topic_switch_points_each ?? 6, 10),
        parseInt(cfg.topic_switch_points_cap ?? 24, 10)
      );
      signals.push({
        code: "topic_switching",
        label: "Unrelated topic changes",
        detail: `${topic_switch_count} topic changes across the archive (threshold ${switchThreshold})`,
        points,
      });
    }
  }

  // --- backlink shape ----------------------------------------------------
  if (total_backlinks !== null && total_backlinks !== undefined && referring_domains) {
    evaluated.push("backlink_shape");
    const ratio = total_backlinks / referring_domains;
    const spikeRatio = parseFloat(cfg.backlink_spike_ratio ?? 3.0);
    // Many links from very few domains is the classic footprint of a
    // sitewide or network link blast.
    if (ratio >= spikeRatio * 25) {
      signals.push({
        code: "backlink_concentration",
        label: "Abnormal backlink concentration",
        detail: `${_fixed0(ratio)} backlinks per referring domain — links come from very few sources`,
        points: parseInt(cfg.backlink_spike_points ?? 14, 10),
      });
    }
  }

  if (lost_backlinks !== null && lost_backlinks !== undefined && total_backlinks) {
    evaluated.push("backlink_decay");
    const lostRatio = lost_backlinks / Math.max(total_backlinks, 1);
    if (lostRatio >= parseFloat(cfg.lost_backlink_ratio ?? 0.6)) {
      signals.push({
        code: "backlink_decay",
        label: "Heavy backlink loss",
        detail: `${_comma(lost_backlinks)} lost vs ${_comma(total_backlinks)} live backlinks (${_fixed0(lostRatio * 100)}%)`,
        points: parseInt(cfg.lost_backlink_points ?? 8, 10),
      });
    }
  }

  if (evaluated.length === 0) {
    // No history, no anchors, no backlinks: we genuinely cannot say.
    return {
      score: null,
      level: null,
      signals: [],
      detected_categories: [],
      evaluated_rules: [],
      has_data: false,
    };
  }

  const score = Math.min(
    100,
    signals.reduce((sum, s) => sum + s.points, 0)
  );
  return {
    score,
    level: levelFor(score),
    signals,
    detected_categories: categories,
    evaluated_rules: evaluated,
    has_data: true,
  };
}

module.exports = {
  LEVEL_LOW,
  LEVEL_MODERATE,
  LEVEL_HIGH,
  LEVEL_VERY_HIGH,
  levelFor,
  spamKeywordSet,
  detectCategories,
  assess,
};
